notebooks/USE_embeddings.py

```diff
@@ -0,0 +1,220 @@
+
+pickle_file = open(os.path.join('data', 'relationship_weights.bin'), 'rb')
+relationship_weights = pickle.load(pickle_file)
+
+# find out whether the weights are directional
+relationship_weights[1]
+relationship_weights_dict = {}
+cid_count = {}
+for i, r in enumerate(relationship_weights):
+    relationship_weights_dict[(r['m.contentID'], r['n.contentID'])] = r['r.weight']
+    cid_count[r['m.contentID']] = cid_count.get(r['m.contentID'], 0) + 1
+    cid_count[r['n.contentID']] = cid_count.get(r['n.contentID'], 0) + 1
+
+# only 22.3k cids with a relationship to other cids
+len(cid_count)
+
+# looking at the counts for various pairs its clear that the counts are directional
+count_1 = relationship_weights_dict[('0008c1e3-c14d-4307-9122-315c50339e49', '96709198-7d36-4cb4-88bc-006890f2c173')]
+count_2 = relationship_weights_dict[('96709198-7d36-4cb4-88bc-006890f2c173', '0008c1e3-c14d-4307-9122-315c50339e49')]
+assert count_1 == count_2, "not the same count"
+
+
+def embed(input):
+    return model(input)
+
+
+# import data from the content store, an improvement here would be to download the same from neo4j
+content_store_df = pd.read_csv("/home/james/Downloads/preprocessed_content_store_210920.csv",
+                               compression='gzip', delimiter="\t", low_memory=False)
+cid_df = pd.DataFrame({'content_id': list(cid_count.keys()), 'relationship_flag': 1})
+cols_keep = ['document_type', 'content_id', 'first_published_at', 'details', 'description']
+subset_content_df = content_store_df.merge(cid_df, on='content_id', how='inner')[cols_keep].copy()
+subset_content_df['details'] = subset_content_df['details'].map(literal_eval)
+
+
+def clean_xml(original_text):
+    ''' strips out xml tagging from string'''
+    extracted_sentence = []
+    start_idx = 1
+    end_idx = 1
+    while (start_idx > 0) and (end_idx > 0):
+        end_idx = original_text.find('<')
+        if end_idx >= 0:
+            extracted_sentence.append(original_text[:end_idx])
+            start_idx = original_text.find('>')
+            if (start_idx >= 0):
+                original_text = original_text[start_idx + 1:]
+    if len(original_text) > 0:
+        extracted_sentence.append(original_text)
+    return str(''.join(extracted_sentence))
+
+
+def extract_paragraphs(original_text):
+    ''' takes raw string text from gov uk and returns extracted paragraphs
+    still contains xml tags'''
+    extracted_paragraphs = []
+    start_idx = 1
+    end_idx = 1
+    while (start_idx >= 0) and (end_idx >= 0):
+        start_idx = original_text.find('<p>')
+        end_idx = original_text.find('</p>')
+        if (start_idx >= 0) and (end_idx >= 0):
+            if (end_idx - start_idx) > 3:
+                cleaned_text_segment = clean_xml(original_text[start_idx + 3:end_idx])
+                extracted_paragraphs.append(cleaned_text_segment)
+            original_text = original_text[end_idx + 3:]
+    return extracted_paragraphs
```

Comment:
One feels that they can take advantage of `BeautifulSoup` to extract the paragraphs via the `<p>` tags but have to contend with taking out the *body* part only.